File: platforms/swift/Sources/ShopifyCheckoutKit/RetryAfter.swift
import Foundation

enum RetryAfter {
    private static let dateFormats = [
        "EEE, dd MMM yyyy HH:mm:ss zzz",
        "EEEE, dd-MMM-yy HH:mm:ss zzz",
        "EEE MMM d HH:mm:ss yyyy"
    ]

    static func seconds(from response: HTTPURLResponse, now: Date = Date()) -> TimeInterval? {
        seconds(from: response.value(forHTTPHeaderField: "Retry-After"), now: now)
    }

    static func seconds(from value: String?, now: Date = Date()) -> TimeInterval? {
        guard let value = value?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else {
            return nil
        }

        if let delay = TimeInterval(value), delay >= 0 {
            return delay
        }

        for format in dateFormats {
            let formatter = DateFormatter()
            formatter.calendar = Calendar(identifier: .gregorian)
            formatter.locale = Locale(identifier: "en_US_POSIX")
            formatter.timeZone = TimeZone(secondsFromGMT: 0)
            formatter.dateFormat = format
            formatter.isLenient = false

            if let date = formatter.date(from: value) {
                return ceil(max(0, date.timeIntervalSince(now)))
            }
        }

        return nil
    }
}

File: platforms/android/lib/src/main/java/com/shopify/checkoutkit/RetryAfter.kt
package com.shopify.checkoutkit

import java.text.ParsePosition
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import kotlin.math.ceil

internal object RetryAfter {
    private const val MILLISECONDS_PER_SECOND = 1_000.0
    private val dateFormats = listOf(
        "EEE, dd MMM yyyy HH:mm:ss zzz",
        "EEEE, dd-MMM-yy HH:mm:ss zzz",
        "EEE MMM d HH:mm:ss yyyy",
    )

    fun seconds(value: String?, nowMillis: Long = System.currentTimeMillis()): Long? {
        val normalized = value?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        val delay = normalized.toLongOrNull()?.takeIf { it >= 0 }
        return delay ?: secondsFromDate(normalized, nowMillis)
    }

    private fun secondsFromDate(value: String, nowMillis: Long): Long? =
        dateFormats.firstNotNullOfOrNull { format ->
            val formatter = SimpleDateFormat(format, Locale.US).apply {
                calendar = Calendar.getInstance(TimeZone.getTimeZone("GMT"), Locale.US)
                timeZone = TimeZone.getTimeZone("GMT")
                isLenient = false
            }
            val position = ParsePosition(0)
            formatter.parse(value, position)
                ?.takeIf { position.index == value.length }
                ?.let { ceil(maxOf(0L, it.time - nowMillis) / MILLISECONDS_PER_SECOND).toLong() }
        }
}
